rasa/shared/core/domain.py

```diff
@@ -229,15 +229,17 @@ def _get_session_config(session_config: Dict) -> SessionConfig:
     @classmethod
     def from_directory(cls, path: Text) -> "Domain":
         """Loads and merges multiple domain files recursively from a directory tree."""
-
-        domain = Domain.empty()
+        domain_dict = {}
         for root, _, files in os.walk(path, followlinks=True):
             for file in files:
                 full_path = os.path.join(root, file)
                 if Domain.is_domain_file(full_path):
-                    other = Domain.from_file(full_path)
-                    domain = other.merge(domain)
-
+                    _ = Domain.from_file(full_path)  # does the validation here only
+                    other_dict = rasa.shared.utils.io.read_yaml(
+                        rasa.shared.utils.io.read_file(full_path)
+                    )
+                    domain_dict = Domain.merge_domain_dicts(domain_dict, other_dict)
+        domain = Domain.from_dict(domain_dict)
         return domain
 
     def merge(self, domain: Optional["Domain"], override: bool = False) -> "Domain":
```

Comment:
I can see that you kept the old `merge` function and that it's used elsewhere. I'm wondering if that's deliberate, especially since the two functions duplicate a lot of code. Maybe the docstrings could clarify the difference between the two?

Reply:
Yes. Unfortunately because of how the `Domain` is loaded, the `merge` function is not aware of all the `entities` it needs to be aware of when there are multiple domain files. I've kept `merge` for now because we're planning an overhaul of the Domain class and fixing this bug has some urgency. I'll clarify in the docstrings.